### ISSUE TITLE
Custom modelform issues

### ISSUE THREAD

**Issue by Hedde** [yourlabs/django-autocomplete-light#440]
I have tested some of my earlier reported issues with the newest branch and I still have problems with my use case

fork: https://github.com/Hedde/django-autocomplete-light (NEW fork)
exampleapp: admin_readonly_issue

Use case:
- I want to add autosuggest to some mandatory fields in the admin
- I want the saved choice to be immutable after a primary key is assigned, so I tried to override get_readonly_fields

When adding a custom ModelForm to the admin with a modelchoice field, there's several thing going wrong, I have added documentation lines to the admin.py file and here's some screenshots;

**saving a modelform with readonly fields**

![schermafbeelding 2015-06-03 om 12 44 44](https://cloud.githubusercontent.com/assets/1533552/7958374/01dec0f2-09ef-11e5-9a18-cb321e4c3123.png)

**saving a modelform with ModelY.x left empty with required=False on the modelform**

![schermafbeelding 2015-06-03 om 12 48 31](https://cloud.githubusercontent.com/assets/1533552/7958377/063f40ea-09ef-11e5-8659-4a0be2c9faff.png)

**Comment by jpic** [comment]
Is it just me or is the field duplicated even when using a
django.forms.ModelChoiceField instead of
autocomplete_light.ModelChoiceField ?
Le 3 juin 2015 11:54, "Hedde van der Heide" <notifications@github.com> a
écrit :

**Comment by jpic** [comment]
Updated above comment.

**Comment by jpic** [comment]
Reproduced here: https://github.com/jpic/double_readonly_field

**Comment by jpic** [comment]
We've got a Trac ticket for this now: https://code.djangoproject.com/ticket/24908#ticket

**Comment by Hedde** [comment]
Sorry I was out yesterday, but you're absolutely right, this seems to be a django issue rather than it has anything to do with autocomplete, I could reproduce your scenario.

**Comment by jpic** [comment]
Could you confirm that this is fixed by https://github.com/django/django/pull/4747/

Thanks B)

**Comment by Hedde** [comment]
Confirmed

**Comment by jpic** [comment]
Awesome, the fix will be backported in autocomplete_light.ModelForm after it's merged into Django 1.8 or master.

**Comment by jpic** [comment]
Related: https://code.djangoproject.com/ticket/8620

**Comment by jpic** [comment]
One fix for current stable django is for you to use get_form to pass a form wiith your_field=None when the obj has no pk, you could use type for that.

**Comment by jpic** [comment]
Fix is on the way for Django: https://code.djangoproject.com/ticket/8620

Your temporary fix:

```
class ModelYAdmin(admin.ModelAdmin):
    def get_readonly_fields(self, request, obj=None):
        # ...
    
    def get_form(self, request, obj=None, **kwargs):
        # Workaround https://code.djangoproject.com/ticket/24908
        if obj:
            self.form = type(self.form.__name__, (self.form,), {'x': None})
        return super(ModelYAdmin, self).get_form(request, obj, **kwargs)
```

Closing, please re-open if you still have any issue. Otherwise, njoy life B)

**Comment by Hedde** [comment]
Thanks, I came up with a similar solution :-) I reckon this will be fixed in 1.8.X, thanks for the immediate action!

**Comment by jpic** [comment]
This should be fixed in 1.8 by https://github.com/django/django/commit/f39d9865cc4f7b8f6ea34a75e45725112a7dc765?diff=unified

Rock'on B)
